### PR TITLE
Initial steps toward a torch.utils.data.IterableDataset (PR 3 of N)

### PR DESCRIPTION
**Pull request by bkmartinjr**
This PR adds a simple `torch.utils.data.IterableDataset` and `torchdata.datapipes.iter.IterDataPipe` implementation.  An initial set of tests is included.

Notes to reviewer:
* multi-worker and multi-GPU support - will partition automatically by both. Detects both Ligtning and `torch.distributed` variants of world and rank.
* API refactored from CZI contribution for better UX:
    * accepts an experiment query directly, removing the need for redundant params
    * encoders removed
    * supports full context config inheritance from the ExperimentAxisQuery allowing access to non-Census datasets
* There is *no* shuffling support - that will come in a later PR
* There is no `dataloader` convenience API - that will come in a later PR

### CODE REVIEW

**Review by ryan-williams — COMMENTED** [review]
Most of the way through, couple q's here, and suggested changes in #15!

File: src/tiledbsoma_ml/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,797 @@
...
+
+        Lifecycle:
+            experimental
+        """
+        self._init_once()
+        assert self._obs_joinids is not None
+        assert self._var_joinids is not None
+        world_size, _ = _get_distributed_world_rank()
+        n_workers, _ = _get_worker_world_rank()
+        partition_len = len(self._obs_joinids) // world_size // n_workers
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply src/tiledbsoma_ml/pytorch.py:349]: Yes, they can vary across data loader workers.  And it doesn't really matter.  What matters the most is having the same number of samples distributed to each GPU.  Beyond that, things are data and config dependent.

File: src/tiledbsoma_ml/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,797 @@
...
+        div, rem = divmod(partition_len, self.batch_size)
+        return div + bool(rem), len(self._var_joinids)
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply src/tiledbsoma_ml/pytorch.py:351]: It is the convention as defined by the data loader framework. They basically want the number of batches (yields of the iterator) on the first dimension.  Hence the mode switching.

Keep in mind that this space is full of awkward conventions, which are not entirely the same across different execution environments (Lightning, etc).  So I went with the conventions inherited from the execution contexts, at least as much as I was able to find a middle path :-)

File: tests/test_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+import pandas as pd
+import pyarrow as pa
+import pytest
+import tiledbsoma as soma
+from scipy import sparse
+from scipy.sparse import coo_matrix, spmatrix
+from tiledbsoma import Experiment, _factory
+from tiledbsoma._collection import CollectionBase
+
+# Conditionally import torch, as it will not be available in all test environments.
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply tests/test_pytorch.py:22]: Excellent point - this _used_ to be true when the code was part of tiledbsoma.  It is no longer true, so I'll push a cleanup momentarily.

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply tests/test_pytorch.py]: Also removed the extraneous `print()` that was in the test code.

**Review by ryan-williams — APPROVED** [review]
🎉
